### ISSUE TITLE
Successful file validation results in empty session data

### ISSUE THREAD

**Issue by pedzed** [laravel/framework#13217]
I am not sure how to describe the problem. To reproduce, add these files to a fresh Laravel (5.2) installation:

**app/Http/routes.php**
```php
<?php

Route::get('/', [
    'as' => 'home',
    'uses' => 'HomeController@index',
]);

Route::post('/', [
    'uses' => 'HomeController@postIndex',
]);
```

**app/Http/Controller/HomeController.php**
```php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Validator;

class HomeController extends Controller
{
    public function index()
    {
        return view('welcome');
    }
    
    public function postIndex(Request $request)
    {
        $validation = Validator::make($request->all(), [
            'input_1' => 'required',
            'file_upload' => 'required',
            'multi_file_upload' => 'required',
            'multi_file_upload.*' => [
                'max:1000000',
                'mimes:zip,x-rar-compressed',
            ],
        ]);
        
        return redirect()
            ->route('home')
            ->withInput($request->all())
            ->withErrors($validation->messages())
        ;
    }
}
```

**resources/views/welcome.blade.php**
```php
<!DOCTYPE html>
<html>
    <head>
        <title>Laravel</title>

        <link href="https://fonts.googleapis.com/css?family=Roboto:100" rel="stylesheet" type="text/css">

        <style>
            html, body {
                height: 100%;
            }

            body {
                margin: 0;
                padding: 0;
                width: 100%;
                display: table;
                font-weight: 100;
                font-family: 'Roboto';
            }

            .container {
                text-align: center;
                display: table-cell;
                vertical-align: middle;
            }

            .content {
                text-align: left;
                display: inline-block;
            }

            .title {
                font-size: 24px;
            }
            
            label {
                display: block;
                margin-bottom: 12px;
            }
        </style>
    </head>
    <body>
        <div class="container">
            <div class="content">
                <h1 class="title">
                    Laravel upload bug
                </h1>
                
                @foreach($errors->all() as $error)
                    <li>{{ $error }}</li>
                @endforeach
                
                <form method="POST"
                      enctype="multipart/form-data"
                      novalidate
                >
                    {!! csrf_field() !!}
                    
                    <label>
                        Example Input:
                        <input type="text"
                               name="input_1"
                               value="{{ old('input_1') }}"
                               placeholder="Lorem ipsum" 
                        />
                    </label>
                    
                    <label>
                        File upload:
                        <input type="file"
                               name="file_upload"
                        />
                    </label>
                    
                    <label>
                        Multi file upload:
                        <input type="file"
                               name="multi_file_upload[]"
                               multiple
                        />
                    </label>
                    
                    <button type="submit">
                        Submit
                    </button>
                </form>
            </div>
        </div>
    </body>
</html>
```

Sorry for the lack of information, I just can't wrap my head about the problem.

*Tested on Linux.*

**Comment by pedzed** [comment]
The problem seems to be partially caused by this line:
```php
name="multi_file_upload[]"
```
If you remove the brackets (`[]`) and change the validation to the following:
```php
$validation = Validator::make($request->all(), [
    'input_1' => 'required',
    'file_upload' => 'required',
    'multi_file_upload' => [
        'required',
        'max:1000000',
        'mimes:zip,x-rar-compressed',
    ],
]);
```
Then it works slightly better. `multi_file_uploads` seems to work, but `file_upload` is bugged in that case.

**Comment by GrahamCampbell** [comment]
Thank you, but this is a duplicate. We are aware of bugs with large files coming in.

**Comment by pedzed** [comment]
> We are aware of bugs with large files coming in.

That was another issue I had, but I increased the values in `php.ini`. This issue is not about that, but about validating multiple files.

As mentioned in my second comment, removing `[]` will make the validation work, but then it will only take the last selected file.

Forget about large files. How are we supposed to validate multiple files in the first place?

**Comment by GrahamCampbell** [comment]
Hmmm, not sure.

**Comment by themsaid** [comment]
I'm not sure I see the problem you are trying to pinpoint, here's my test:

```php
Route::post('/', function () {
        $validation = Validator::make(Request::all(), [
            'input_1' => 'required',
            'file_upload' => 'required',
            'multi_file_upload' => 'required',
            'multi_file_upload.*' => [
                'max:1',
                'mimes:zip,x-rar-compressed',
            ],
        ]);

        dd($validation->messages());
    });
```

And the rules seem to be working:

<img width="346" alt="screen shot 2016-04-19 at 10 44 20 pm" src="https://cloud.githubusercontent.com/assets/4332182/14654919/3819225a-0680-11e6-9ef3-1caa666ed9a2.png">

**Comment by pedzed** [comment]
@themsaid What's your operating system? I really hope this is Linux-specific, because it should be an easy fix then.

Any possibility you could test this on Linux?

**Comment by themsaid** [comment]
I'm on Ubuntu 14.04.1, hmm not sure what might be wrong.

**Comment by pedzed** [comment]
@themsaid Could you tell me whether you get `$files` and the old input in the view if you pass the rules?

```php
$validation = Validator::make($request->all(), [
    'input_1' => 'required',
    'file_upload' => [
        'max:1000',
        'mimes:zip,x-rar-compressed',
    ],
    'multi_file_upload.*' => [
        'max:1000000',
        'mimes:zip,x-rar-compressed',
    ],
]);

if($validation->fails()) {
    session()->flash('flash_message', 'Validation failed!');
} else {
    session()->flash('flash_message', 'Validation succeeded!');
}

$files = [];

if($request->hasFile('multi_file_upload')) {
    foreach($request->file('multi_file_upload') as $file) {
        $files[] = $file;
        // \App\Upload::create([
        //     'name' => $file->getClientOriginalName(),
        // ]);
    }
}

return redirect()
    ->route('home')
    ->with('files', $files)
    ->withInput($request->all())
    ->withErrors($validation->messages())
;
```
The session seems to be reset...

**Comment by themsaid** [comment]
Is your issue with validation failing or something else?

**Comment by pedzed** [comment]
> Is your issue with validation failing or something else?

If validation fails, then everything works fine. The session gets saved. But if the validation passes, I believe the session gets reset.

**Comment by pedzed** [comment]
@themsaid Can you reproduce the problem?

**Comment by ctf0** [comment]
@pedzed i believe what u r after is u want to validate an array of files so the rules should be as
```php
 $validation = Validator::make($request->all(), [
            'input_1' => 'required',
            'file_upload' => 'required',
            'multi_file_upload.*' => [
                'required_without:file_upload',
                'max:1000000',
                'mimes:zip,x-rar-compressed',
            ],
        ]);
```
and in the view u can retrieve the errors by ```{{$errors->get('multi_file_upload.0')}} but this is buggy from version to version 'ex.v5.2.36` which displaying the errors as above doesn't work, but was downgraded back to `v5.2.31` so update ur application and see if it works.

**Comment by pedzed** [comment]
I updated Laravel to v5.2.35 and the problem still exists. I hope you understand the problem, because I myself found it difficult to understand it.

The problem is that the session data gets emptied under certain circumstances. You lose flash sessions, old input, errors, etc.

**Comment by ctf0** [comment]
@pedzed try `$errors->all()` this will work, the second one wont work in the latest release and here is an issue that describe that but no answers yet https://github.com/laravel/framework/issues/13779

**Comment by themsaid** [comment]
Closing since it appears to be fixed in 5.3, please feel free to ping me if you still face the problem.
